lib/fluent/plugin/out_google_cloud.rb

```diff
@@ -873,15 +641,409 @@ def fetch_gce_metadata(metadata_path)
+        type: group_resource_type,
+        labels: group_resource_labels.to_h
+      )
+
+      # Freeze the per-request state. Any further changes must be made on a
+      # per-entry basis.
+      group_resource.freeze
+      group_resource.labels.freeze
+      group_common_labels.freeze
+
+      [group_resource, group_common_labels]
+    end
+
+    # Determine group level monitored resource type shared by a collection of
+    # entries.
+    # Return the resource type and tag regexp matched groups. The matched groups
+    # only apply to some resource types. Return nil if not applicable or if
+    # there is no match.
+    def determine_group_level_monitored_resource_type(tag)
+      @tag_regexp_list.each do |rule_pair|
+        matched_regex_group = rule_pair[:tag_regexp].match(tag)
+        return [rule_pair[:derived_type], matched_regex_group] if
+          matched_regex_group
+      end
+      [@resource.type, nil]
+    end
+
+    # Determine group level monitored resource labels and common labels. These
+    # labels will be shared by a collection of entries.
+    def determine_group_level_labels(group_resource_type, matched_regex_group)
+      group_resource_labels = @resource.labels.dup
+      group_common_labels = @common_labels.dup
+
+      case group_resource_type
+      # Cloud Functions.
+      when CLOUDFUNCTIONS_CONSTANTS[:resource_type]
+        group_resource_labels.merge!(
+          'region' => @gcf_region,
+          'function_name' => decode_cloudfunctions_function_name(
+            matched_regex_group['encoded_function_name'])
+        )
+
+        instance_id = group_resource_labels.delete('instance_id')
+        group_common_labels.merge!(
+          "#{CONTAINER_CONSTANTS[:service]}/instance_id" => instance_id,
+          "#{COMPUTE_CONSTANTS[:service]}/resource_id" => instance_id,
+          "#{CONTAINER_CONSTANTS[:service]}/cluster_name" =>
+            group_resource_labels.delete('cluster_name'),
+          "#{COMPUTE_CONSTANTS[:service]}/zone" =>
+            group_resource_labels.delete('zone')
+        )
+
+      # GKE container.
+      when CONTAINER_CONSTANTS[:resource_type]
+        unless matched_regex_group.nil?
+          matched_regex_group_hash = Hash[
+            matched_regex_group.names.zip(matched_regex_group.captures)]
+          group_resource_labels.merge!(
+            delete_and_extract_labels(
+              matched_regex_group_hash, {
+                # The kubernetes_tag_regexp is poorly named. 'namespace_name' is
+                # in fact 'namespace_id'. 'pod_name' is in fact 'pod_id'.
+                'container_name' => 'container_name',
+                'namespace_name' => 'namespace_id',
+                'pod_name' => 'pod_id'
+              }, false))
+
+          group_common_labels.merge!(
+            delete_and_extract_labels(
+              matched_regex_group_hash,
+              CONTAINER_CONSTANTS[:extra_common_labels]
```

Comment:
[Optional] I find `Hash[list.map { ... }]` easier to read than `list.map { ... }.to_h`, because the hash construction is upfront. Also in other places below.